File: jcv-core/src/main/java/com/ekino/oss/jcv/core/JsonComparator.java
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import com.ekino.oss.jcv.core.validator.ValidatorTemplateManager;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.skyscreamer.jsonassert.Customization;
import org.skyscreamer.jsonassert.JSONCompareMode;
import org.skyscreamer.jsonassert.JSONCompareResult;
import org.skyscreamer.jsonassert.ValueMatcher;
import org.skyscreamer.jsonassert.ValueMatcherException;
import org.skyscreamer.jsonassert.comparator.DefaultComparator;
import org.skyscreamer.jsonassert.comparator.JSONComparator;

import static java.util.stream.Collectors.*;
import static org.skyscreamer.jsonassert.comparator.JSONCompareUtil.*;

/**
 * Custom {@link JSONComparator}.
 *
 * @author Leo Millon
 */
public class JsonComparator extends DefaultComparator {

    private static final String IGNORED_PATH = "";

    private final List<JsonValidator> validators;

    public JsonComparator(JSONCompareMode mode, List<JsonValidator> validators) {
        super(mode);
        this.validators = new LinkedList<>(validators);
    }

    public List<JsonValidator> getValidators() {
        return Collections.unmodifiableList(validators);
    }

    @Override
    public void compareValues(String prefix, Object expectedValue, Object actualValue, JSONCompareResult result) throws JSONException {

        Optional<Customization> matchingValueCustomization = validators.stream()
            .filter(it -> it.getContextMatcher().matches(prefix, expectedValue, actualValue))
            .findFirst()
            .map(toMatcher())
            .map(it -> new Customization(IGNORED_PATH, it));

        if (matchingValueCustomization.isPresent()) {
            try {
                if (!matchingValueCustomization.get().matches(prefix, actualValue, expectedValue, result)) {
                    result.fail(prefix, expectedValue, actualValue);
                }
            } catch (ValueMatcherException e) {
                result.fail(prefix, e);
            }
            return;
        }

        super.compareValues(prefix, expectedValue, actualValue, result);
    }

    @SuppressWarnings("unchecked")
    private static Function<JsonValidator, ValueMatcher<Object>> toMatcher() {
        return it -> (ValueMatcher<Object>) it.getValueComparator();
    }

    @Override
    protected void compareJSONArrayOfJsonObjects(String key, JSONArray expected, JSONArray actual, JSONCompareResult result) throws JSONException {
        String uniqueKey = findUniqueKey(expected);
        if (uniqueKey == null || !isUsableAsUniqueKey(uniqueKey, actual)) {
            // An expensive last resort
            recursivelyCompareJSONArray(key, expected, actual, result);
            return;
        }
        Map<Object, JSONObject> expectedValueMap = arrayOfJsonObjectToMap(expected, uniqueKey);
        Map<Object, JSONObject> actualValueMap = arrayOfJsonObjectToMap(actual, uniqueKey);
        for (Object id : expectedValueMap.keySet()) {
            if (!actualValueMap.containsKey(id)) {
                result.missing(formatUniqueKey(key, uniqueKey, id), expectedValueMap.get(id));
                continue;
            }
            JSONObject expectedValue = expectedValueMap.get(id);
            JSONObject actualValue = actualValueMap.get(id);
            compareValues(formatUniqueKey(key, uniqueKey, id), expectedValue, actualValue, result);
        }
        for (Object id : actualValueMap.keySet()) {
            if (!expectedValueMap.containsKey(id)) {
                result.unexpected(formatUniqueKey(key, uniqueKey, id), actualValueMap.get(id));
            }
        }
    }

    /**
     * @see org.skyscreamer.jsonassert.comparator.JSONCompareUtil#findUniqueKey(JSONArray)
     */
    private static String findUniqueKey(JSONArray expected) throws JSONException {
        // Find a unique key for the object (id, name, whatever)
        JSONObject o = (JSONObject) expected.get(0); // There's at least one at this point
        for (String candidate : getKeys(o)) {
            if (isUsableAsUniqueKey(candidate, expected)) return candidate;
        }
        // No usable unique key :-(
        return null;
    }

    /**
     * @see org.skyscreamer.jsonassert.comparator.JSONCompareUtil#isUsableAsUniqueKey(String, JSONArray)
     */
    private static boolean isUsableAsUniqueKey(String candidate, JSONArray array) throws JSONException {
        Collection<Object> seenValues = new HashSet<>();
        for (int i = 0; i < array.length(); i++) {
            Object item = array.get(i);
            if (item instanceof JSONObject) {
                JSONObject o = (JSONObject) item;
                if (o.has(candidate)) {
                    Object value = o.get(candidate);
                    // rewrite original code to evict validator templates from valid unique key
                    if (isSimpleValue(value) && !seenValues.contains(value) && !new ValidatorTemplateManager(value.toString()).isValidTemplate()) {
                        seenValues.add(value);
                    } else {
                        return false;
                    }
                } else {
                    return false;
                }
            } else {
                return false;
            }
        }
        return true;
    }

    @Override
    protected void compareJSONArrayOfSimpleValues(String key, JSONArray expected, JSONArray actual, JSONCompareResult result) throws JSONException {
        List<Object> expectedElements = jsonArrayToList(expected);
        List<Object> actualElements = jsonArrayToList(actual);

        List<ExpectedElement> parsedExpectedElements = parseExpectedElements(key, expectedElements);

        if (parsedExpectedElements.stream().noneMatch(ExpectedElement::hasCustomization)) {
            super.compareJSONArrayOfSimpleValues(key, expected, actual, result);
            return;
        }

        Set<Integer> actualValueMatchedIndexes = new HashSet<>();
        Map<ExpectedElement, Optional<ActualElement>> matchingByValue = getMatchingByValue(parsedExpectedElements, actualElements, actualValueMatchedIndexes);
        Map<ExpectedElement, Collection<ActualElement>> matchingByValidator = getExpectedElementCollectionMap(parsedExpectedElements, key, actualElements, actualValueMatchedIndexes);

        long totalMatched = matchingByValue.values().stream().filter(Optional::isPresent).count()
            + matchingByValidator.values().stream().flatMap(Collection::stream).distinct().count();

        if (totalMatched != actualElements.size()) {

            Map<ExpectedElement, Collection<ActualElement>> allMatches = Stream.concat(
                matchingByValue.entrySet().stream()
                    .map(entry -> new AbstractMap.SimpleEntry<ExpectedElement, Collection<ActualElement>>(entry.getKey(), entry.getValue().map(Collections::singletonList).orElseGet(Collections::emptyList)))
                    .map(it -> (Map.Entry<ExpectedElement, Collection<ActualElement>>) it),
                matchingByValidator.entrySet().stream()
            )
                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));

            Set<Integer> allMatchedActualIndexes = allMatches.values().stream()
                .flatMap(Collection::stream)
                .map(ActualElement::getIndex)
                .collect(toSet());

            IntStream.range(0, actualElements.size())
                .filter(it -> !allMatchedActualIndexes.contains(it))
                .forEachOrdered(actualIndex -> result.unexpected(key + "[" + actualIndex + "]", actualElements.get(actualIndex)));

            String detailedMatchingDebugMessage = allMatches
                .entrySet()
                .stream()
                .sorted(Comparator.comparing(entry -> entry.getKey().getIndex()))
                .map(entry -> {
                    ExpectedElement expectedElt = entry.getKey();
                    String matchedElements = entry.getValue().stream()
                        .map(it -> "[" + it.getIndex() + "] -> " + it.getValue())
                        .collect(joining(",", "[", "]"));
                    return key + "[" + expectedElt.getIndex() + "] -> " + expectedElt.getKey() + " matched with: " + matchedElements;
                })
                .collect(joining("\n"));
            result.fail(detailedMatchingDebugMessage);
        }
    }

    private List<ExpectedElement> parseExpectedElements(String key, List<Object> expectedElements) {
        List<ExpectedElement> parsedExpectedElements = new ArrayList<>();
        for (int i = 0; i < expectedElements.size(); i++) {
            Object expectedElement = expectedElements.get(i);
            parsedExpectedElements.add(new ExpectedElement(
                i,
                expectedElement,
                validators.stream()
                    .filter(it -> it.getContextMatcher().matches(key, expectedElement, null))
                    .findFirst()
                    .map(toMatcher())
                    .map(it -> new Customization(IGNORED_PATH, it))
                    .orElse(null)
            ));
        }
        return parsedExpectedElements;
    }

    private Map<ExpectedElement, Optional<ActualElement>> getMatchingByValue(List<ExpectedElement> parsedExpectedElements,
                                                                             List<Object> actualElements,
                                                                             Set<Integer> actualValueMatchedIndexes) {
        return parsedExpectedElements.stream()
            .filter(it -> !it.hasCustomization())
            .map(expectedElement -> {
                for (int i = 0; i < actualElements.size(); i++) {
                    if (actualValueMatchedIndexes.contains(i)) {
                        continue;
                    }
                    Object actualElement = actualElements.get(i);
                    if (expectedElement.getKey().equals(actualElement)) {
                        actualValueMatchedIndexes.add(i);
                        return new AbstractMap.SimpleEntry<>(expectedElement, Optional.of(new ActualElement(i, actualElement)));
                    }
                }
                return new AbstractMap.SimpleEntry<>(expectedElement, Optional.<ActualElement>empty());
            })
            .collect(toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
    }

    private Map<ExpectedElement, Collection<ActualElement>> getExpectedElementCollectionMap(List<ExpectedElement> parsedExpectedElements,
                                                                                            String key,
                                                                                            List<Object> actualElements,
                                                                                            Set<Integer> actualValueMatchedIndexes) {
        return parsedExpectedElements.stream()
            .filter(ExpectedElement::hasCustomization)
            .map(expectedElement -> {
                Set<ActualElement> matched = new HashSet<>();
                for (int i = 0; i < actualElements.size(); i++) {
                    if (actualValueMatchedIndexes.contains(i)) {
                        continue;
                    }
                    Object actualElement = actualElements.get(i);
                    try {
                        if (expectedElement.getCustomization().matches(key, actualElement, expectedElement.getKey(), new JSONCompareResult())) {
                            matched.add(new ActualElement(i, actualElement));
                        }
                    } catch (ValueMatcherException e) {
                        // Do nothing
                    }
                }
                return new AbstractMap.SimpleEntry<>(expectedElement, matched);
            })
            .collect(toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
    }

    private static class ExpectedElement {
        private final Integer index;
        private final Object key;
        private final Customization customization;

        private ExpectedElement(Integer index, Object key, Customization customization) {
            this.index = index;
            this.key = key;
            this.customization = customization;
        }

        Integer getIndex() {
            return index;
        }

        Object getKey() {
            return key;
        }

        Customization getCustomization() {
            return customization;
        }

        boolean hasCustomization() {
            return customization != null;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ExpectedElement that = (ExpectedElement) o;
            return Objects.equals(index, that.index) &&
                Objects.equals(key, that.key);
        }

        @Override
        public int hashCode() {
            return Objects.hash(index, key);
        }
    }

    private static class ActualElement {
        private final Integer index;
        private final Object value;

        ActualElement(Integer index, Object value) {
            this.index = index;
            this.value = value;
        }

        Integer getIndex() {
            return index;
        }

        Object getValue() {
            return value;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ActualElement that = (ActualElement) o;
            return Objects.equals(index, that.index) &&
                Objects.equals(value, that.value);
        }

        @Override
        public int hashCode() {
            return Objects.hash(index, value);
        }
    }
}

File: jcv-core/src/main/kotlin/com/ekino/oss/jcv/core/JsonComparator.kt
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core

import com.ekino.oss.jcv.core.validator.ValidatorTemplateManager
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.skyscreamer.jsonassert.Customization
import org.skyscreamer.jsonassert.JSONCompareMode
import org.skyscreamer.jsonassert.JSONCompareResult
import org.skyscreamer.jsonassert.ValueMatcher
import org.skyscreamer.jsonassert.ValueMatcherException
import org.skyscreamer.jsonassert.comparator.DefaultComparator
import org.skyscreamer.jsonassert.comparator.JSONComparator
import org.skyscreamer.jsonassert.comparator.JSONCompareUtil.arrayOfJsonObjectToMap
import org.skyscreamer.jsonassert.comparator.JSONCompareUtil.formatUniqueKey
import org.skyscreamer.jsonassert.comparator.JSONCompareUtil.getKeys
import org.skyscreamer.jsonassert.comparator.JSONCompareUtil.isSimpleValue
import org.skyscreamer.jsonassert.comparator.JSONCompareUtil.jsonArrayToList

/**
 * Custom [JSONComparator].
 *
 * @author Leo Millon
 */
class JsonComparator(mode: JSONCompareMode, validators: List<JsonValidator<out Any?>>) : DefaultComparator(mode) {

    private val validators: List<JsonValidator<out Any?>> = validators.toList()

    @Throws(JSONException::class)
    override fun compareValues(prefix: String, expectedValue: Any?, actualValue: Any?, result: JSONCompareResult) {

        validators
            .firstOrNull { it.contextMatcher.matches(prefix, expectedValue, actualValue) }
            ?.let { asCustomization(it) }
            ?.let {
                try {
                    if (!it.matches(prefix, actualValue, expectedValue, result)) {
                        result.fail(prefix, expectedValue, actualValue)
                    }
                } catch (e: ValueMatcherException) {
                    result.fail(prefix, e)
                }
                return
            }
            ?: super.compareValues(prefix, expectedValue, actualValue, result)
    }

    @Suppress("UNCHECKED_CAST")
    private fun asCustomization(it: JsonValidator<out Any?>) =
        Customization("", it.valueComparator as ValueMatcher<Any>?)

    @Throws(JSONException::class)
    override fun compareJSONArrayOfJsonObjects(
        key: String,
        expected: JSONArray,
        actual: JSONArray,
        result: JSONCompareResult
    ) {
        val uniqueKey = findUniqueKey(expected)
        if (uniqueKey == null || !isUsableAsUniqueKey(uniqueKey, actual)) {
            // An expensive last resort
            recursivelyCompareJSONArray(key, expected, actual, result)
            return
        }

        val expectedValueMap = arrayOfJsonObjectToMap(expected, uniqueKey)
        val actualValueMap = arrayOfJsonObjectToMap(actual, uniqueKey)

        expectedValueMap
            .asSequence()
            .forEach { (expectedId, expectedValue) ->
                if (!actualValueMap.containsKey(expectedId)) {
                    result.missing(formatUniqueKey(key, uniqueKey, expectedId), expectedValue)
                    return@forEach
                }
                val actualValue = actualValueMap[expectedId]
                compareValues(formatUniqueKey(key, uniqueKey, expectedId), expectedValue, actualValue, result)
            }

        actualValueMap
            .asSequence()
            .filterNot { (actualId, _) -> expectedValueMap.containsKey(actualId) }
            .forEach { (actualId, actualValue) ->
                result.unexpected(formatUniqueKey(key, uniqueKey, actualId), actualValue)
            }
    }

    @Throws(JSONException::class)
    override fun compareJSONArrayOfSimpleValues(
        key: String,
        expected: JSONArray,
        actual: JSONArray,
        result: JSONCompareResult
    ) {
        val expectedElements = jsonArrayToList(expected)
        val actualElements = jsonArrayToList(actual)

        val parsedExpectedElements = parseExpectedElements(key, expectedElements)

        if (parsedExpectedElements.none { it.hasCustomization() }) {
            super.compareJSONArrayOfSimpleValues(key, expected, actual, result)
            return
        }

        val actualValueMatchedIndexes = mutableSetOf<Int>()
        val matchingByValue = getMatchingByValue(parsedExpectedElements, actualElements, actualValueMatchedIndexes)
        val matchingByValidator =
            getExpectedElementCollectionMap(parsedExpectedElements, key, actualElements, actualValueMatchedIndexes)

        val totalMatched =
            matchingByValue.values.asSequence().filterNotNull().count() + matchingByValidator.values.asSequence().flatten().distinct().count()

        if (totalMatched == actualElements.size) {
            return
        }

        val allMatches = sequenceOf(
            matchingByValue.asSequence()
                .map { (key, value) -> key to value?.let { listOf(it) }.orEmpty() },
            matchingByValidator.asSequence().map { it.toPair() }
        )
            .flatten()
            .toMap()

        val allMatchedActualIndexes = allMatches.values.asSequence()
            .flatten()
            .mapNotNull { it.index }
            .toSet()

        actualElements
            .forEachIndexed { index, actualElement ->
                actualElement
                    .takeUnless { allMatchedActualIndexes.contains(index) }
                    ?.also { result.unexpected("$key[$index]", it) }
            }

        val detailedMatchingDebugMessage = allMatches
            .asSequence()
            .sortedBy { it.key.index }
            .map { (expectedElt, value) ->
                val matchedElements = value
                    .joinToString(",", "[", "]") { "[" + it.index + "] -> " + it.value }
                key + "[" + expectedElt.index + "] -> " + expectedElt.key + " matched with: " + matchedElements
            }
            .joinToString("\n")

        result.fail(detailedMatchingDebugMessage)
    }

    private fun parseExpectedElements(key: String, expectedElements: List<Any>): List<ExpectedElement> {
        return expectedElements
            .asSequence()
            .mapIndexed { index, expectedElement ->
                validators
                    .firstOrNull { it.contextMatcher.matches(key, expectedElement, null) }
                    ?.let { asCustomization(it) }
                    .let { ExpectedElement(index, expectedElement, it) }
            }
            .toList()
    }

    private fun getMatchingByValue(
        parsedExpectedElements: List<ExpectedElement>,
        actualElements: List<Any>,
        actualValueMatchedIndexes: MutableSet<Int>
    ): Map<ExpectedElement, ActualElement?> {
        return parsedExpectedElements
            .asSequence()
            .filterNot { it.hasCustomization() }
            .associateWith { expectedElement ->
                actualElements
                    .asSequence()
                    .mapIndexedNotNull { index, actualElement ->
                        actualElement
                            .takeUnless { actualValueMatchedIndexes.contains(index) }
                            ?.takeIf { it -> expectedElement.key == it }
                            ?.also { actualValueMatchedIndexes.add(index) }
                            ?.let { ActualElement(index, actualElement) }
                    }
                    .firstOrNull()
            }
    }

    private fun getExpectedElementCollectionMap(
        parsedExpectedElements: List<ExpectedElement>,
        key: String?,
        actualElements: List<Any>,
        actualValueMatchedIndexes: Set<Int>
    ): Map<ExpectedElement, Collection<ActualElement>> {
        return parsedExpectedElements
            .asSequence()
            .filter { it.hasCustomization() }
            .associateWith { expectedElement ->
                actualElements
                    .asSequence()
                    .mapIndexedNotNull { index, actualElement ->
                        actualElement
                            .takeUnless { actualValueMatchedIndexes.contains(index) }
                            ?.takeIf { it ->
                                try {
                                    expectedElement.customization
                                        ?.matches(key, it, expectedElement.key, JSONCompareResult())
                                        ?: false
                                } catch (e: ValueMatcherException) {
                                    false
                                }
                            }
                            ?.let { ActualElement(index, it) }
                    }
                    .toSet()
            }
    }

    private data class ExpectedElement(val index: Int?, val key: Any, val customization: Customization?) {

        fun hasCustomization(): Boolean {
            return customization != null
        }
    }

    private data class ActualElement(val index: Int?, val value: Any)

    /**
     * @see org.skyscreamer.jsonassert.comparator.JSONCompareUtil.findUniqueKey
     */
    @Throws(JSONException::class)
    private fun findUniqueKey(expected: JSONArray): String? {
        // Find a unique key for the object (id, name, whatever)
        return getKeys((expected[0] as JSONObject))
            .firstOrNull { isUsableAsUniqueKey(it, expected) } // if null, no usable unique key :-(
    }

    /**
     * @see org.skyscreamer.jsonassert.comparator.JSONCompareUtil.isUsableAsUniqueKey
     */
    @Throws(JSONException::class)
    private fun isUsableAsUniqueKey(candidate: String, array: JSONArray): Boolean {
        val seenValues = mutableSetOf<Any>()
        for (i in 0 until array.length()) {
            array[i]
                .takeIfInstanceOf<JSONObject>()
                ?.takeIf { it.has(candidate) }
                ?.let { it[candidate] }
                ?.takeIf { isSimpleValue(it) && !seenValues.contains(it) && !ValidatorTemplateManager(it.toString()).isValidTemplate }
                ?.also { seenValues.add(it) }
                ?: return false
        }
        return true
    }

    private inline fun <reified T> Any?.takeIfInstanceOf(): T? {
        return if (this is T) this else null
    }
}
